skfp/utils/splitters.py

```diff
@@ -0,0 +1,221 @@
+    """
+    np.testing.assert_almost_equal(train_size + test_size, 1.0)
+
+    scaffolds = create_scaffolds(list(data), include_chirality)
+    scaffold_sets = sorted(scaffolds.values(), key=len)
+
+    train_cutoff: int = int(train_size * len(data))
+    train_ids: list[int] = []
+    test_ids: list[int] = []
+
+    for scaffold_set in scaffold_sets:
+        if len(train_ids) < train_cutoff:
+            train_ids.extend(scaffold_set)
+        else:
+            test_ids.extend(scaffold_set)
+
+    if return_indices:
+        return train_ids, test_ids
+    else:
+        return _get_data_from_indices(
+            data, scaffolds, train_ids
+        ), _get_data_from_indices(data, scaffolds, test_ids)
+
+
+def scaffold_train_valid_test_split(
+    data: Sequence[Union[str, Mol]],
+    train_size: float = 0.8,
+    test_size: float = 0.1,
+    valid_size: float = 0.1,
+    return_indices: bool = False,
+    include_chirality: bool = False,
+) -> tuple[list, list, list]:
+    """
+    Split a list of SMILES or RDKit Mol objects into train, validation,
+    and test subsets using Bemis-Murcko scaffolds.
+
+    This function ensures that similar molecules (sharing the same scaffold) are in the same split.
+    This method, suggested in the MoleculeNet paper, helps to test the model's ability to generalize
+    to entirely new scaffolds.
+
+    Parameters
+    ----------
+    data : list[str]
+        Sequence representing either SMILES strings or RDKit Mol objects.
+
+    frac_train : float, default=0.8
+        The fraction of data to be used for the training split.
+
+    frac_valid : float, default=0.1
+        The fraction of data to be used for the validation split.
+
+    frac_test : float, default=0.1
+        The fraction of data to be used for the test split.
+
+    return_indices : bool, default=False
+        Should the method return SMILES/RDKit Mol objects or just indices in the dataset.
+
+    include_chirality: bool, default=False
+        Should the method take chirality of molecules into consideration?
+
+
+    Returns
+    ----------
+    subsets : tuple[list, list, list]
+        A tuple of train, validation, and test subsets. The format depends on `return_indices`:
+        - If `return_indices` is False, returns lists of SMILES strings
+        or RDKit Mol objects depending on the input.
+        - If `return_indices` is True, returns lists of indices.
+
+    Notes
+    -----
+    - This implementation uses Bemis-Murcko scaffolds (Bemis and Murcko, 1996) to group molecules.
+    - The split is fully deterministic, with the smallest scaffold sets assigned to the test subset,
+    larger sets to the validation subset,
+      and the largest sets to the training subset.
+    - The test and validation sets can be slightly larger than specified,
+    ensuring they are at least the desired fraction, but as close as possible in size.
+    - The split fractions (frac_train, frac_valid, frac_test) must sum to 1.
```

Comment:
Make sure that those notes are identical for both train-test and train-valid-test functions